design/multi-factor-admin-and-step-up.md

```diff
@@ -0,0 +1,613 @@
+
+```
+Auth type
+------------------------------------------------------------
+Identity Provider Redirector
+Authentication factor (1)
+  |- Cookie
+  |- Kerberos
+  |- Username Password Form
+Authentication factor (2) [OPTIONAL]
+  |- Cookie
+  |- Kerberos
+  |- OTP Form
+```
+And if we wanted to have an mandatory 2nd factor, with either OTP or a FIDO2 configured:
+
+```
+Auth type
+------------------------------------------------------------
+Identity Provider Redirector
+Authentication factor (1)
+  |- Cookie
+  |- Kerberos
+  |- Username Password Form
+Authentication factor (2)
+  |- Cookie
+  |- Kerberos
+  |- OTP Form
+  |- FIDO-2
+```
+Potentially we could also introduce another type of authentication execution, we could call it "Authenticated on", to simplify the authenticators that bypass the authentication factors. We could then have:
+
+```
+Auth type
+------------------------------------------------------------
+Identity Provider Redirector
+Authenticated on
+  |- Cookie
+  |- Kerberos
+Authentication factor (1)
+  |- Username Password Form
+Authentication factor (2)
+  |- OTP Form
+  |- FIDO-2
+```
+
+#### Authentication categories
+
+Authentication categories allow the link between the actual execution within the flow and the
+credentials of a certain type.
+
+For example, currently there is the OTPFormAuthenticator for the browser flow and ValidateOTP for the direct grant flow.
+However, both these authentication executions will accept input from the same OTP credentials. This means that schematically, we would have:
+```
+-----------------------         ----------------------
+| OTP Credential 1    |---------|                    |    ------------------------
+-----------------------         |                    |----| OTPFormAuthenticator |
+-----------------------         | OTP Authentication |    ------------------------
+| OTP Credential 2    |---------|      Category      |    ---------------
+-----------------------         |                    |----| ValidateOTP |
+-----------------------         |                    |    ---------------
+| OTP Credential 3    |---------|                    |
+-----------------------         ----------------------
+```
+
+This schema serves as a base for all credentials and allows them to be
+used for the appropriate authentication executions. This link allows Keycloak to present
+alternative credentials for an execution during login.
+
+To simplify selection in the GUI, it would also be useful to do some refactoring to ensure
+that the execution code for the `Browser` / `Registration Flow` / `Direct Grant Flow` /
+`Reset Credentials` / `Client Authentication` Bindings can only be selected as executions
+for their binding categories. This already seems to be the case for `Client Authentication`,
+and would simplify the selection of authentication executions within a authentication
+factor.
+
+It may also allow a simplification of the naming when selecting an authentication
```

Comment:
Not sure about that. I see your point, but I would rather still keep the naming like "OTP" + "OTP Form" to have it more clear which Authenticator implementation will be used. But maybe it's just me... Not sure what is better regarding usability for admins.

Reply:
If we introduce a way to limit what flow types an authenticator is compatible with Form is no longer needed in the name as only OTP Form would be available in the browser flow.